server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java

```diff
@@ -1089,14 +1089,7 @@ public IndexMetadata build() {
                 throw new IllegalArgumentException("must specify positive number of shards for index [" + index + "]");
             }
 
-            Integer maybeNumberOfReplicas = settings.getAsInt(SETTING_NUMBER_OF_REPLICAS, null);
-            if (maybeNumberOfReplicas == null) {
-                throw new IllegalArgumentException("must specify numberOfReplicas for index [" + index + "]");
-            }
-            int numberOfReplicas = maybeNumberOfReplicas;
-            if (numberOfReplicas < 0) {
-                throw new IllegalArgumentException("must specify non-negative number of replicas for index [" + index + "]");
-            }
+            int numberOfReplicas = INDEX_NUMBER_OF_REPLICAS_SETTING.get(settings);
```

Comment:
This isn't equivalent, since not having this setting set in the metadata is considered illegal.

Reply:
That is, we want to maintain that these are set upstream, and not fallback to defaults here. We want to do that elsewhere, so that if we somehow end up here without it having been set, it's a bug that we ended up with metadata without having it set.

Reply:
There were two cases before. The first is the missing case. It was never intended to end up here without number of replicas set; it isn't actually an error from the user's perspective. The default was previously handled upstream, but now we are using the default mechanism in Setting. There's no reason to set number of replicas just so it can be passed in here. The second case was the bounds check, which is also handled in the Setting by the minimum value of 0.

Reply:
We copy/rebuild index metadata in various places (shrink, split, clone, restores, etc.). This check ensures that when we rebuild the index metadata, that we have set the number of replicas (and also we do this for the number of shards), instead of relying on the defaults.

Reply:
I took at a stab and what I think this should look like in #56801.